NOnion/Network/TorStream.fs

```diff
@@ -142,200 +139,377 @@ type TorStream(circuit: TorCircuit) =
+                                        + (readFromCurrentBuffer
+                                            buffer
+                                            (offset + bytesRead)
+                                            (length - bytesRead))
+
+                                    let newBytesRemaining =
+                                        length - newBytesRead
+
+                                    if incomingCells.Count = 0 then
+                                        return newBytesRead
                                     else
-                                        return bytesRead
-                                }
+                                        return!
+                                            tryRead
+                                                newBytesRead
+                                                newBytesRemaining
+                                else
+                                    return bytesRead
+                            }
 
-                            return! tryRead 0 length
-                }
+                        return! tryRead 0 length
+            }
 
-            return! receiveLock.RunAsyncWithSemaphore safeReceive
+
+        async {
+            let! cancellationToken = Async.CancellationToken
+            cancellationToken.ThrowIfCancellationRequested()
+
+            let! {
+                     StreamBuffer = buffer
+                     BufferOffset = offset
+                     BufferLength = length
+                     ReplyChannel = replyChannel
+                 } = inbox.Receive()
+
+            do!
+                safeReceive(buffer, offset, length)
+                |> TryExecuteAsyncAndReplyAsResult replyChannel
+
+            return! StreamReceiveMailBoxProcessor inbox
+        }
+
+    let streamReceiveMailBox =
+        MailboxProcessor.Start(StreamReceiveMailBoxProcessor)
+
+    static member Accept (streamId: uint16) (circuit: TorCircuit) =
+        async {
+            let stream = TorStream circuit
+            do! stream.RegisterIncomingStream streamId
+
+            do! circuit.SendRelayCell streamId (RelayConnected Array.empty) None
+
+            sprintf
+                "TorStream[%i,%i]: incoming stream accepted"
+                streamId
+                circuit.Id
+            |> TorLogger.Log
+
+            return stream
+        }
+
+    member __.End() =
+        async {
+            let! sendResult =
+                streamControlMailBox.PostAndAsyncReply StreamControlMessage.End
+
+            return UnwrapResult sendResult
+        }
+
+    member self.EndAsync() =
+        self.End() |> Async.StartAsTask
+
+
+    member __.SendData(data: array<byte>) =
+        async {
+            let! sendResult =
+                streamControlMailBox.PostAndAsyncReply(fun replyChannel ->
+                    StreamControlMessage.Send(data, replyChannel)
+                )
+
+            return UnwrapResult sendResult
```

Comment:
@aarani Unwrap without message? let's fix the API of UnwrapResult to have an err msg

Reply:
No unwrap here works by just re-raising the exception that it contains which I think is enough as a msg, isn't it?

Reply:
fair enough